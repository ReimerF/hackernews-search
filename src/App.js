import React, { Component } from "react";
import "./App.css";
import Search from "./components/Search.js";
import Table from "./components/Table.js";
import Button from "./components/Button.js";
import fetch from "isomorphic-fetch";

const list = [];

const DEFAULT_QUERY = "redux";
const DEFAULT_HPP = "100";

const PATH_BASE = "https://hn.algolia.com/api/v1";
const PATH_SEARCH = "/search";
const PARAM_SEARCH = "query=";
const PARAM_PAGE = "page=";
const PARAM_HPP = "hitsPerPage=";

const Loading = () => <div>Loading...</div>;

const withLoading = Component => ({ isLoading, ...rest }) =>
  isLoading ? <Loading /> : <Component {...rest} />;

const ButtonWithLoading = withLoading(Button);

class App extends Component {
  state = {
    results: null,
    searchKey: "",
    list,
    error: null,
    searchTerm: DEFAULT_QUERY,
    page: 0,
    isLoading: false
  };

  setSearchTopStories = result => {
    const { hits, page } = result;
    const { results, searchKey } = this.state;

    const oldHits =
      results && results[searchKey] ? results[searchKey].hits : [];

    const updatedHits = [...oldHits, ...hits];
    const updatedResults = {
      ...results,
      [searchKey]: { hits: updatedHits, page }
    };
    const updatedList = updatedResults[searchKey].hits;

    this.setState({
      ...this.state,
      results: updatedResults,
      list: updatedList,
      isLoading: false
    });
  };

  fetchSearchTopStories = (searchTerm, page = 0) => {
    this.setState({ isLoading: true });
    console.log(
      `fetching ${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}`
    );
    fetch(
      `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`
    )
      .then(response => response.json())
      .then(result => this.setSearchTopStories(result))
      .catch(e => {
        this.setState({ error: e, isLoading: false });
      });
  };

  componentDidMount() {
    const { searchTerm } = this.state;
    this.fetchSearchTopStories(searchTerm);
    this.setState({ searchKey: searchTerm });
  }

  onDismiss = id => {
    console.log("starting onDismiss...");
    const { results, searchKey } = this.state;
    const { hits, page } = results[searchKey];
    console.log("hits", hits);
    const updatedList = hits.filter(item => item.objectID !== id);
    console.log("updated list", updatedList);
    console.log("current list before update", this.state.list);
    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: updatedList, page }
      },
      list: updatedList
    });
    console.log("list after update", this.state.list);
  };

  onSearchChange = textoBusca => {
    this.setState({ searchTerm: textoBusca.target.value });
  };

  onSearchSubmit = event => {
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });
    this.fetchSearchTopStories(searchTerm);
    event.preventDefault();
  };

  render() {
    const {
      searchTerm,
      results,
      list,
      searchKey,
      isLoading,
      error
    } = this.state;
    const page =
      (results && results[searchKey] && results[searchKey].page) || 0;

    return (
      <div className="page">
        <div className="interactions">
          <Search
            searchTerm={searchTerm}
            onSearchSubmit={this.onSearchSubmit}
            onSearchChange={this.onSearchChange}
          >
            {" "}
            Search{" "}
          </Search>
        </div>
        {!error ? (
          isLoading ? (
            <div>Loading...</div>
          ) : (
            <>
              <Table
                list={list}
                searchTerm={searchTerm}
                onDismiss={this.onDismiss}
              />
              <div className="interactions">
                <Button
                  onClick={() =>
                    this.fetchSearchTopStories(searchKey, page + 1)
                  }
                >
                  {" "}
                  More{" "}
                </Button>
              </div>
            </>
          )
        ) : (
          <h3>Something went wrong.</h3>
        )}
      </div>
    );
  }
}

export default App;

export { Button, Search, Table };
