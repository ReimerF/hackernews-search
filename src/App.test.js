import React from "react";
import ReactDOM from "react-dom";
import { render } from "@testing-library/react";
import App, {
  Search,
  Button,
  Table,
  onSearchChange,
  onSearchSubmit,
  searchTerm,
  onClick
} from "./App";
import renderer from "react-test-renderer";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

Enzyme.configure({ adapter: new Adapter() });

// test('renders learn react link', () => {
//   const { getByText } = render(<App />);
//   const linkElement = getByText(/learn react/i);
//   expect(linkElement).toBeInTheDocument();
// });

describe("App", () => {
  it("renders App without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<App />, div);
  });

  test("App has a valid snapshot", () => {
    const component = renderer.create(<App />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Search", () => {
  it("renders Search without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <Search
        onSearchChange={onSearchChange}
        onSearchSubmit={onSearchSubmit}
        searchTerm={searchTerm}
      >
        Search
      </Search>,
      div
    );
  });

  test("Search has a valid snapshot", () => {
    const component = renderer.create(
      <Search
        onSearchChange={onSearchChange}
        onSearchSubmit={onSearchSubmit}
        searchTerm={searchTerm}
      >
        Search
      </Search>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Button", () => {
  it("renders Button without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <Button onClick={onClick}>Testing happening here</Button>,
      div
    );
  });

  test("Button has a valid snapshot", () => {
    const component = renderer.create(
      <Button onClick={onClick}>Testing happening here</Button>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Table", () => {
  const props = {
    list: [
      {
        title: "1",
        author: "2",
        num_comments: "3",
        points: "4",
        objectID: "x"
      },
      { title: "5", author: "6", num_comments: "7", points: "8", objectID: "y" }
    ]
  };

  it("shows two items in list", () => {
    const element = shallow(<Table {...props} />);
    expect(element.find(".table-row").length).toBe(2);
  });

  test("has a valid snapshot", () => {
    const component = renderer.create(<Table {...props} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
