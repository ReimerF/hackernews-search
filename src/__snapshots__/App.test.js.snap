// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App App has a valid snapshot 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <div>
      <form
        onSubmit={[Function]}
      >
        <button
          type="submit"
        >
           
          Search
           
           &gt;
        </button>
        <input
          onChange={[Function]}
          type="text"
          value="redux"
        />
      </form>
    </div>
  </div>
  <div>
    Loading...
  </div>
</div>
`;

exports[`Button Button has a valid snapshot 1`] = `
<button
  type="button"
>
  Testing happening here
</button>
`;

exports[`Search Search has a valid snapshot 1`] = `
<div>
  <form>
    <button
      type="submit"
    >
      Search
       &gt;
    </button>
    <input
      type="text"
    />
  </form>
</div>
`;

exports[`Table has a valid snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a>
        1
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      3
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      4
    </span>
    <button
      className="button-inline"
      onClick={[Function]}
      type="button"
    >
       Dismiss 
    </button>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a>
        5
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      6
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      7
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      8
    </span>
    <button
      className="button-inline"
      onClick={[Function]}
      type="button"
    >
       Dismiss 
    </button>
  </div>
</div>
`;
